new CDSE STAC API and speed up of stacking SEN2 tiles

konstntokas opened the pull request:
This PR contains the following points:

- This PR implements the new CDSE STAC API with the endpoint https://stac.dataspace.copernicus.eu/v1
- Tiles are stacked along the time axis, subsequently reprojected to a common grid, and stitched together (mosaicking) where first value is taken. This procedure can be used after https://github.com/xcube-dev/xcube/pull/1098, which allows to perform `rectify_dataset` on a 3D dataset. This reduces the number of reprojection and therefore speeds up the stacking/mosaicking process.

forman commented on xcube_stac/constants.py:
```diff
@@ -144,34 +141,80 @@
-    description="Names of assets (bands) which will be included in the data cube.",
+    description="Names of assets which will be included in the data cube.",
+)
+SCEMA_APPLY_SCALING = JsonBooleanSchema(
```
```suggestion
SCHEMA_APPLY_SCALING = JsonBooleanSchema(
```

forman commented on xcube_stac/constants.py:
```diff
@@ -144,34 +141,80 @@
-    apply_scaling=JsonBooleanSchema(
-        title="Apply scaling, offset, and no-data values to data"
-    ),
+    apply_scaling=SCEMA_APPLY_SCALING,
```
```suggestion
    apply_scaling=SCHEMA_APPLY_SCALING,
```

forman commented on xcube_stac/stack.py:
```diff
@@ -84,4 +134,29 @@ def mosaic_take_first(list_ds: list[xr.Dataset]) -> xr.Dataset:
         )
     if "crs" in list_ds[0]:
         ds_mosaic["crs"] = list_ds[0].crs
+    if "spatial_ref" in list_ds[0]:
+        ds_mosaic.coords["spatial_ref"] = list_ds[0].spatial_ref
```
Remember that `crs` and `spatial_ref` are not names that are put in stone. These can vary. The meaning of the grid mapping variable only comes from referring it from other (data) variables using attribute `grid_mapping`. You can also detect grid mapping variables by their own required attribute `grid_mapping_name`.

konstntokas replied:
I already addressed https://github.com/xcube-dev/xcube/issues/1013 for xcube stac. I added a function `normalize_grid_mapping`, which is only called once at the end once.

forman commented on xcube_stac/stack.py:
```diff
@@ -84,4 +134,29 @@ def mosaic_take_first(list_ds: list[xr.Dataset]) -> xr.Dataset:
+        final_slices.append(ds_mosaic)
+    final_ds = xr.concat(final_slices, dim="time", join="exact")
+    final_ds = final_ds.assign_coords(coords=dict(time=dts))
+    if "crs" in final_ds:
```
See above.

konstntokas replied:
see above.

forman commented on xcube_stac/store_mode.py:
```diff
@@ -484,15 +481,21 @@ def open_data(
             )
         return ds
 
     def stack_items(
```
The method's logic is not easy grasp, hence its body would benefit from explaining comments or refactoring out methods with expressive names.

konstntokas replied:
I added a comment.

forman commented on xcube_stac/helper.py:
```diff
@@ -134,153 +120,48 @@ def get_data_access_params(self, item: pystac.Item, **open_params) -> dict:
-                    f"A{item.properties["orbitNumber"]:06}_{time_end}/IMG_DATA/"
-                    f"R{res_select}m/T{item.properties["tileId"]}_"
-                    f"{id_parts[2]}_{asset_name}_{res_select}m.jp2"
+    def list_assets_from_item(
```
Please consider putting sentinel-2 (/CDSE) related logic into a separate module.

konstntokas replied:
I will do so in the next PR

forman commented on xcube_stac/constants.py:
```diff
@@ -144,34 +141,80 @@
     spatial_res=SCHEMA_SPATIAL_RES,
     query=SCHEMA_ADDITIONAL_QUERY,
 )
+
```
Please condier moving Sentinel-2 related stuff into separate module.

konstntokas replied:
I will do so in the next PR.

forman commented on xcube_stac/store.py:
```diff
@@ -262,17 +268,8 @@ def __init__(
+        self._format_ids = ["zarr", "levels"]
 
 
 class StacCdseDataStore(StacDataStore):
```
Move into separate submodule.

forman reviewed (changes requested):
Looks good overall. You removed the `sen2` module, but actually I'd prefer we had a separate `sen2` module that contains sentinel-2 specific logic.

konstntokas commented:
> Looks good overall. You removed the `sen2` module, but actually I'd prefer we had a separate `sen2` module that contains sentinel-2 specific logic.

I felt the same regarding the `sen2` module, after I implemented more Sentinel-2 specific functionalities for reading out the angles. In the next PR there will be a package containing all accessors, and I implemented a module for the https, s3 and sen2 accessor.

konstntokas commented on xcube_stac/_utils.py:
```diff
@@ -218,6 +144,16 @@ def search_collections(
+    else:
+        format_id = MAP_MIME_TYP_FORMAT.get(asset.media_type.split("; ")[0])
+    if format_id is None:
+        LOG.debug(f"No format_id found for asset {asset.title!r}")
```
I removed it.

konstntokas commented on xcube_stac/_utils.py:
```diff
@@ -364,13 +300,20 @@ def add_nominal_datetime(items: list[pystac.Item]) -> list[pystac.Item]:
+    return float(
+        item.properties.get(
+            "processing:version",
+            item.properties.get("s2:processing_baseline", "1.0"),
```
the defaults it not S2 specific. If no processing version is given, then 1.0 returned. We need a string which can be turned into a float, because the code looks for the highest processing version. I suggest to move it to the stack module.

forman reviewed (approved):
Please see my minor suggestions. Or just run black.
Since we use Python 3.10 consider using `|` instead of `Union` and optionally other goodies from `typing` such `TypeAlias`. Next time maybe.

konstntokas commented:
> Please see my minor suggestions. Or just run black. Since we use Python 3.10 consider using `|` instead of `Union` and optionally other goodies from `typing` such `TypeAlias`. Next time maybe.

I changed it using `pyupgrade`, which closes #16

konstntokas commented on xcube_stac/accessor.py:
```diff
@@ -155,18 +152,13 @@ def open_data(
                 (
-                    f"{access_params["protocol"]}://{access_params["root"]}/"
-                    f"{access_params["fs_path"]}"
+                    f"{access_params['protocol']}://{access_params['root']}/"
+                    f"{access_params['fs_path']}"
```
Done
